specs/simple-serialize.md

```diff
@@ -1,424 +1,108 @@
-2. Prefix the concatenation with its length encoded as a `4-byte` **little-endian** unsigned integer.
-
-We define `bytes` to be a synonym of `List[bytes1]`.
-
-**Example in Python**
+#### Containers, tuples, lists
 
 ```python
-
-serialized_list_string = b''
-
-for item in value:
-   serialized_list_string += serialize(item)
-
-assert(len(serialized_list_string) < 2**32)
-
-serialized_len = (len(serialized_list_string).to_bytes(LENGTH_BYTES, 'little'))
-
-return serialized_len + serialized_list_string
+serialized_bytes = ''.join([serialize(element) for element in value])
+LENGTH_BYTES = 4
+assert len(serialized_bytes) < 2**(8 * LENGTH_BYTES)
+serialized_length = len(serialized_bytes).to_bytes(LENGTH_BYTES, 'little')
+return serialized_length + serialized_bytes
 ```
 
-#### Container
+## Deserialization
 
-A container represents a heterogenous, associative collection of key-value pairs. Each pair is referred to as a `field`. To get the value for a given field, you supply the key which is a symbol unique to the container referred to as the field's `name`. The container data type is analogous to the `struct` type found in many languages like C or Go.
+Given a type, serialization is an injective function from objects of that type to byte strings. That is, deserialization—the inverse function—is well-defined.
 
-To serialize a container, obtain the list of its field's names in the specified order. For each field name in this list, obtain the corresponding value and serialize it. Tightly pack the complete set of serialized values in the same order as the field names into a buffer. Calculate the size of this buffer of serialized bytes and encode as a `4-byte` **little endian** `uint32`. Prepend the encoded length to the buffer. The result of this concatenation is the final serialized value of the container.
+## Merkleization
 
-| Check to perform                              | Code                        |
-|:----------------------------------------------|:----------------------------|
-| Length of serialized fields fits into 4 bytes | ``len(serialized) < 2**32`` |
+We first define helper functions:
 
-To serialize:
-
-1. Get the list of the container's fields.
-
-2. For each name in the list, obtain the corresponding value from the container and serialize it. Place this serialized value into a buffer. The serialized values should be tightly packed.
-
-3. Get the number of raw bytes in the serialized buffer. Encode that number as a `4-byte` **little endian** `uint32`.
-
-4. Prepend the length to the serialized buffer.
-
-**Example in Python**
-
-```python
-def get_field_names(typ):
-    return typ.fields.keys()
+* `pack`: Given ordered objects of the same basic type, serialize them, pack them into 32-byte chunks, right-pad the last chunk with zero bytes, and return the chunks.
+* `merkleize`: Given ordered 32-byte chunks, right-pad them with zero chunks to the next power of two, Merkleize the chunks, and return the root.
```

Comment:
Make it clearer that the number of chunks should be a power of two

Reply:
How do you suggest changing "right-pad them with zero chunks to the next power of two"?

Reply:
Reworded to "if necessary append zero chunks so that the number of chunks is a power of two" 👍